logstash-core/spec/logstash/instrument/metric_store_spec.rb

```diff
@@ -0,0 +1,103 @@
+
+    before do
+      subject.fetch_or_store(namespaces, key, counter)
+    end
+
+    it "return the object" do
+      expect(subject.fetch_or_store(namespaces, key, new_counter)).to eq(counter)
+    end
+  end
+
+  context "when the namespace end node isn't a map" do
+    let(:conflicting_namespaces) { [:root, :pipelines, :pipeline_01, :events_in] }
+
+    before do
+      subject.fetch_or_store(namespaces, key, counter)
+    end
+
+    it "raise an exception" do
+      expect { subject.fetch_or_store(conflicting_namespaces, :new_key, counter) }.to raise_error(LogStash::Instrument::MetricStore::ConcurrentMapExpectedError)
+    end
+  end
+
+  describe "#get" do
+    let(:metric_events) {
+      [
+        [[:node, :sashimi, :pipelines, :pipeline01, :plugins, :"logstash-output-elasticsearch"], :event_in, :increment],
+        [[:node, :sashimi, :pipelines, :pipeline01, :plugins, :"logstash-output-elasticsearch"], :event_in, :increment],
+        [[:node, :sashimi, :pipelines, :pipeline01], :processed_events_in, :increment],
+        [[:node, :sashimi, :pipelines, :pipeline01], :processed_events_out, :increment],
+      ]
+    }
+
+    before do
+      # Lets add a few metrics in the store before trying to convert them
+      metric_events.each do |namespaces, metric_key, action|
+        metric = subject.fetch_or_store(namespaces, metric_key, LogStash::Instrument::MetricType::Counter.new(namespaces, key))
+        metric.execute(action)
+      end
+    end
+
+    it "retrieves end of branch values" do
+      metrics = subject.get(:node, :sashimi, :pipelines, :pipeline01, :plugins, :"logstash-output-elasticsearch")
+      expect(metrics.size).to eq(1)
+    end
+
+    it "retrieves branch" do
+      metrics = subject.get(:node, :sashimi, :pipelines, :pipeline01)
+      expect(metrics.size).to eq(3)
+    end
+
+    it "allow to retrieve a specific metrics" do
+      metrics = subject.get(:node, :sashimi, :pipelines, :pipeline01, :plugins, :"logstash-output-elasticsearch", :event_in)
+      expect(metrics).to be_kind_of(LogStash::Instrument::MetricType::Base)
+    end
+  end
+
+  describe "#to_event" do
+    let(:metric_events) {
+      [
+        [[:node, :sashimi, :pipelines, :pipeline01, :plugins, :"logstash-output-elasticsearch"], :event_in, :increment],
+        [[:node, :sashimi, :pipelines, :pipeline01, :plugins, :"logstash-output-elasticsearch"], :event_in, :increment],
+        [[:node, :sashimi, :pipelines, :pipeline01], :processed_events, :increment],
+      ]
+    }
+
+    before do
+      # Lets add a few metrics in the store before trying to convert them
+      metric_events.each do |namespaces, metric_key, action|
+        metric = subject.fetch_or_store(namespaces, metric_key, LogStash::Instrument::MetricType::Counter.new(namespaces, key))
+        metric.execute(action)
+      end
+    end
+
+    it "converts all metric to `Logstash::Event`" do
+      events = subject.to_events
+      expect(events.size).to eq(2)
```

Comment:
you should add an specific test for this. I see this being a pattern in your test here, you include a duplicated metrics event, and test some deduplication like his. I think is better to include a separated case as this is very important behaviour and here for example is mix with the event conversion.